core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,413 @@
+	}
+	priorityFeeThresholdWei := assets.NewWei(priorityFeeThreshold)
+
+	u.priorityFeeThresholdMu.Lock()
+	u.priorityFeeThreshold = priorityFeeThresholdWei
+	u.priorityFeeThresholdMu.Unlock()
+
+	// eth_feeHistory may return less results than BlockHistorySize so we need to divide by the length of the result
+	maxPriorityFeePerGas := assets.NewWei(priorityFee.Div(priorityFee, big.NewInt(int64(len(feeHistory.Reward)))))
+	// baseFeeBufferPercentage is used as a safety to catch fluctuations in the next block.
+	maxFeePerGas := baseFee.AddPercentage(BaseFeeBufferPercentage).Add((maxPriorityFeePerGas))
+
+	promUniversalEstimatorBaseFee.WithLabelValues(u.chainID.String()).Set(float64(baseFee.Int64()))
+	promUniversalEstimatorMaxPriorityFeePerGas.WithLabelValues(u.chainID.String()).Set(float64(maxPriorityFeePerGas.Int64()))
+	promUniversalEstimatorMaxFeePerGas.WithLabelValues(u.chainID.String()).Set(float64(maxFeePerGas.Int64()))
+
+	u.logger.Debugf("Fetched new dynamic prices, nextBlock#: %v - oldestBlock#: %v - maxFeePerGas: %v - maxPriorityFeePerGas: %v - maxPriorityFeeThreshold: %v",
+		nextBlock, feeHistory.OldestBlock, maxFeePerGas, maxPriorityFeePerGas, priorityFeeThresholdWei)
+
+	u.dynamicPriceMu.Lock()
+	defer u.dynamicPriceMu.Unlock()
+	u.dynamicPrice.FeeCap = maxFeePerGas
+	u.dynamicPrice.TipCap = maxPriorityFeePerGas
+	return u.dynamicPrice, nil
+}
+
+func (u *UniversalEstimator) getDynamicPrice() (fee DynamicFee, err error) {
+	u.dynamicPriceMu.RLock()
+	defer u.dynamicPriceMu.RUnlock()
+	if u.dynamicPrice.FeeCap == nil || u.dynamicPrice.TipCap == nil {
+		return fee, fmt.Errorf("dynamic price not set")
+	}
+	return u.dynamicPrice, nil
+}
+
+// BumpLegacyGas provides a bumped gas price value by bumping the previous one by BumpPercent.
+// If the original value is higher than the max price it returns an error as there is no room for bumping.
+// It aggregates the market, bumped, and max gas price to provide a correct value.
+func (u *UniversalEstimator) BumpLegacyGas(ctx context.Context, originalGasPrice *assets.Wei, gasLimit uint64, maxPrice *assets.Wei, _ []EvmPriorAttempt) (*assets.Wei, uint64, error) {
+	// Sanitize original fee input
+	if originalGasPrice == nil || originalGasPrice.Cmp(maxPrice) >= 0 {
+		return nil, 0, fmt.Errorf("%w: error while retrieving original gas price: originalGasPrice: %s. Maximum price configured: %s",
+			commonfee.ErrBump, originalGasPrice, maxPrice)
+	}
+
+	currentGasPrice, err := u.getGasPrice()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	bumpedGasPrice := originalGasPrice.AddPercentage(u.config.BumpPercent)
+	bumpedGasPrice, err = LimitBumpedFee(originalGasPrice, currentGasPrice, bumpedGasPrice, maxPrice)
+	if err != nil {
+		return nil, 0, fmt.Errorf("gas price error: %s", err.Error())
+	}
+
+	u.logger.Debugw("bumped gas price", "originalGasPrice", originalGasPrice, "bumpedGasPrice", bumpedGasPrice)
+
+	return bumpedGasPrice, gasLimit, nil
```

Comment:
why return gasLimit when it was passed in to this function? I don't think the value was modified in the function.

Reply:
Unfortunately, this is a byproduct of the `EvmEstimator` interface that the Universal Estimator is implementing, but hopefully, it will change once we deprecate Arbitrum Estimator.